File: packages/contracts-python/anvilkit_contracts/compatibility.py
"""Candidate BOM verification at the consumer boundary."""

from __future__ import annotations

from typing import Any

from .identity import contract_bom


class CompatibilityError(ValueError):
    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code


def verify_candidate_bom(value: dict[str, Any], consumer_generation: int) -> str:
    _, digest, verified = contract_bom(value)
    if not verified:
        raise CompatibilityError("BOM_DIGEST_MISMATCH", "candidate BOM identity does not verify")
    window = value["compatibility"]
    if not window["minimumConsumerGeneration"] <= consumer_generation <= window["maximumConsumerGeneration"]:
        raise CompatibilityError("CONTRACT_UNSUPPORTED", "consumer generation is outside the BOM compatibility window")
    return digest

File: packages/contracts-go/compatibility/compatibility.go
// Package compatibility verifies a candidate BOM at the consumer boundary.
package compatibility

import (
	"encoding/json"
	"fmt"

	contractidentity "github.com/ancyloce/anvilkit-platform/packages/contracts-go/identity"
)

type Error struct {
	Code string
	Text string
}

func (e *Error) Error() string { return e.Text }

// VerifyCandidate verifies the BOM identity and its declared consumer window.
func VerifyCandidate(raw []byte, consumerGeneration int) (string, error) {
	var bom struct {
		Compatibility struct {
			Minimum int `json:"minimumConsumerGeneration"`
			Maximum int `json:"maximumConsumerGeneration"`
		} `json:"compatibility"`
	}
	if err := json.Unmarshal(raw, &bom); err != nil {
		return "", fmt.Errorf("decode candidate BOM: %w", err)
	}
	_, digest, verified, err := contractidentity.ContractBOM(raw)
	if err != nil {
		return "", err
	}
	if !verified {
		return "", &Error{Code: "BOM_DIGEST_MISMATCH", Text: "candidate BOM identity does not verify"}
	}
	if consumerGeneration < bom.Compatibility.Minimum || consumerGeneration > bom.Compatibility.Maximum {
		return "", &Error{Code: "CONTRACT_UNSUPPORTED", Text: "consumer generation is outside the BOM compatibility window"}
	}
	return digest, nil
}
